algebird-core/src/main/scala/com/twitter/algebird/Windows.scala

```diff
@@ -1,142 +1,99 @@
 package com.twitter.algebird
 
-import scala.annotation.tailrec
-
-//          800 900      700
-//          |   |   1000 |
-// Vector ( 19, 20, 23, 16 )
-//    index=2        ^
-//    time=1000
-//    step=100
-
-// [700, 800): saw a sum of 16
-// [800, 900): saw a sum of 19
-// [900, 1000): saw a sum of 20
-// [1000, 1100): saw at least a sum of 23 *in progress*
-
-// 24 hour moving window, with 30 minute resolution (step)
-// (48 + 1 buckets)
-
-// time range you can support: step * (slots - 1)
-case class Windows[A](slots: Vector[A], index: Int, step: Long, time: Long) { lhs =>
-
-  // oldest-to-newest iterator across partial sums
-  def iterator: Iterator[A] =
-    (1 to slots.size).iterator.map(i => slots(makeIndex(index + i)))
-
-  @inline private[this] def makeIndex(i: Long): Int =
-    ((i + slots.size) % slots.size).toInt
-
-  private[this] def clear(vect: Vector[A], i: Int)(implicit ev: Monoid[A]): Vector[A] =
-    vect.updated(i, ev.zero)
-
+/**
+ *         800 900      700
+ *          |   |   1000 |
+ * Vector ( 19, 20, 23, 16 )
+ *    index=2        ^
+ *    time=1000
+ *    step=100
+ *
+ * [700, 800): saw a sum of 16
+ * [800, 900): saw a sum of 19
+ * [900, 1000): saw a sum of 20
+ * [1000, 1100): saw at least a sum of 23 *in progress*
+ *
+ * 24 hour moving window, with 30 minute resolution (step)
+ * (48 + 1 buckets)
+ *
+ * time range you can support: step * (slots - 1)
+ *
+ * We only bump `time` in increments of `step`. `time`
+ * references the inclusive lower bound of the bucket you're currently
+ * in.
+ */
+case class Windows[A](buf: RingBuf[A], step: Long, time: Long) { lhs =>
+  // Inclusive lower bound of the supplied slot offset, looking back
+  // from the current slot.
+  private def minTime: Long = time - (buf.size - 1) * step
+
+  private[this] def stepsFrom(beginning: Long, now: Long): Long =
```

Comment:
this was super useful, and allowed me to kill the tail recursive loop and a lot of code below.